viewing_party/party.py

```diff
@@ -1,23 +1,197 @@
+    return user_data
+
+
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    if not user_data["watched"]:
+        return 0.0
+    
+    total_score = 0
+    movie_num = 0
+    for movie in user_data["watched"]:
+        total_score += movie["rating"]
+        movie_num += 1
+    return total_score / movie_num
+
+def get_most_watched_genre(user_data):
+    watched = user_data["watched"]
+
+    if not watched:
+        return None
+    
+    genre_counts = {}
+    for movie in watched:
+        genre = movie.get("genre")
+        if genre not in genre_counts:
+            genre_counts[genre] = 1
+        else:
+            genre_counts[genre] += 1
+
+    # most_watched_genre = max(genre_counts, key=genre_counts.get)
+
+    max_count = 0
+    
+    for genre, count in  genre_counts.items():
+
+        if count > max_count:
+
+            max_count = count
+            max_watched_genre = genre
+
+
+    return max_watched_genre
 
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+def get_unique_watched(user_data):
+    if not user_data:
+        return None
+    
+    user_watched_movies = user_data["watched"]
+    friends_watched_movies = []  
+
+    for friend in user_data["friends"]:
+        friends_watched_movies.extend(friend["watched"])
+
+    unique_watched = []
+
+    for movie in user_watched_movies:
+        if movie not in friends_watched_movies:
+            unique_watched.append(movie)
+
+    return unique_watched
+
+
+def get_friends_unique_watched(user_data):
+    if not user_data:
+        return None
+    
+    user_watched_movies = user_data["watched"]
+    friends_watched_movies = []  
+
+    for friend in user_data["friends"]:
+        friends_watched_movies.extend(friend["watched"])
+
+    unique_watched = []
+
+    for movie in friends_watched_movies:
+        if movie not in user_watched_movies and movie not in unique_watched:
+            unique_watched.append(movie)
+    
+    return unique_watched
+
+
+
+
+
+
         
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    
+    user_watched = user_data["watched"]
+    subscriptions = user_data["subscriptions"]
+    
+    friend_watched =[]
+    
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie not in friend_watched:
+                friend_watched.append(movie)
+                
+    recs = []
+    for movie in friend_watched:
+       
+        if movie in user_watched:
+            continue
+
+       
+        if movie["host"] not in subscriptions:
+            continue
+
+       
+        recs.append(movie)
+    return recs

```

Comment:
It appears that one of the main ways this function works is by making a list of all the movies our friends have watched that we haven't... I wonder if there is an opportunity to use a function we have already written to simplify our approach here?